internal/component/prometheus/remote/queue/filequeue/filequeue.go

```diff
@@ -0,0 +1,190 @@
+	}
+
+	// Save the existing files in `q.existingFiles`, which will be push to `out` when actor starts.
+	for _, id := range ids {
+		name := filepath.Join(directory, fmt.Sprintf("%d.committed", id))
+		q.existingsFiles = append(q.existingsFiles, name)
+	}
+	return q, nil
+}
+
+func (q *queue) Start() {
+	// Actors and mailboxes have to be started. It makes sense to combine them into one unit since they
+	// have the same lifespan.
+	q.self = actor.Combine(actor.New(q), q.dataQueue).Build()
+	q.self.Start()
+}
+
+func (q *queue) Stop() {
+	q.self.Stop()
+}
+
+// Store will add records to the dataQueue that will add the data to the filesystem. This is an unbuffered channel.
+// Its possible in the future we would want to make it a buffer of 1, but so far it hasnt been an issue in testing.
+func (q *queue) Store(ctx context.Context, meta map[string]string, data []byte) error {
+	return q.dataQueue.Send(ctx, types.Data{
+		Meta: meta,
+		Data: data,
+	})
+}
+
+// get returns the data of the file or an error if something wrong went on.
+func get(logger log.Logger, name string) (map[string]string, []byte, error) {
+	defer deleteFile(logger, name)
+	buf, err := readFile(name)
+	if err != nil {
+		return nil, nil, err
+	}
+	r := &Record{}
+	_, err = r.UnmarshalMsg(buf)
+	if err != nil {
+		return nil, nil, err
+	}
+	return r.Meta, r.Data, nil
+}
+
+// DoWork allows most of the queue to be single threaded with work only coming in and going out via mailboxes(channels).
+func (q *queue) DoWork(ctx actor.Context) actor.WorkerStatus {
+	// Queue up our existing items.
+	for _, name := range q.existingsFiles {
+		q.out(ctx, types.DataHandle{
+			Name: name,
+			Get: func() (map[string]string, []byte, error) {
+				return get(q.logger, name)
+			},
+		})
+	}
+	// We only want to process existing files once.
+	q.existingsFiles = nil
+	select {
+	case <-ctx.Done():
+		return actor.WorkerEnd
+	case item, ok := <-q.dataQueue.ReceiveC():
+		if !ok {
+			return actor.WorkerEnd
+		}
+		name, err := q.add(item.Meta, item.Data)
+		if err != nil {
+			level.Error(q.logger).Log("msg", "error adding item - dropping data", "err", err)
+			return actor.WorkerContinue
+		}
+		// The idea is that this will callee will block/process until the callee is ready for another file.
+		q.out(ctx, types.DataHandle{
+			Name: name,
+			Get: func() (map[string]string, []byte, error) {
+				return get(q.logger, name)
+			},
+		})
+		return actor.WorkerContinue
+	}
+}
+
+// Add a file to the queue (as committed).
+func (q *queue) add(meta map[string]string, data []byte) (string, error) {
+	if meta == nil {
+		meta = make(map[string]string)
+	}
+	q.maxID++
+	name := filepath.Join(q.directory, fmt.Sprintf("%d.committed", q.maxID))
+	r := &Record{
+		Meta: meta,
+		Data: data,
+	}
+	// Not reusing a buffer here since allocs are not bad here and we are trying to reduce memory.
+	rBuf, err := r.MarshalMsg(nil)
+	if err != nil {
+		return "", err
+	}
+	err = q.writeFile(name, rBuf)
```

Comment:
One file = one piece of data implies that we probably want to have a good level of batching for this queue and probably it's worth calling out that it's going to work better with larger data.

Reply:
In general yes, in the case of a single node exporter self scraping its kilobytes. The default is it writes `10,000` signals in the full PR with a flush timer to ensure its timely. When testing internally the files after snappy compressed were megabytes.

Reply:
Yeah, that likely will work if you write entire scrapes to the queue and I'm fine with that... if you started to write single samples then it's gonna be a lot of overhead. That's all fine as long as we make it somewhat clear that the queue is optimised only for some use patterns.

BTW, I'd like to see some metrics on how the queue performs, but I'm happy for this to be in future PRs.

Reply:
There are some e2e benchmarks in the full pr. The file queue is really cheap though.